### ISSUE TITLE
Is schema enforced on outputs and internal data-nodes?

### ISSUE THREAD

**Issue by ankostis** [cedricleroy/pyungo#33]
Adapting the quickstart example:
```python
schema = {
    "type": "object",
    "properties": {
        "a": {"type": "number"},
        "b": {"type": "number"}
    }
}

graph = Graph(schema=schema)

@graph.register(inputs=['a'], outputs=['b'])
def f1(a):
    return "Hey!"

@graph.register(inputs=['b'], outputs=['c'])
def f2(b):
    return 2 * b

graph.calculate(data={'a': 1})
```

I was expecting an error, but got the pipeline went through and got the result `Hey@!Hey!`.
Am i doing something wrong?
This feature is particularly important for data on the internal nodes, because it is not as easy to test them as inputs/outputs.

**Comment by cedricleroy** [comment]
This won't work as the check happens on inputs data. We could have a flag to check outputs as well. Or you can put a contract on `b`:

```python

schema = {
    "type": "object",
    "properties": {
        "a": {"type": "number"},
    }
}

graph = Graph(schema=schema)

@graph.register(inputs=['a'], outputs=Output(name='b', contract='float'))
def f1(a):
    return "Hey!"

@graph.register(inputs=['b'], outputs=['c'])
def f2(b):
    return 2 * b

graph.calculate(data={'a': 1})

```

**Comment by ankostis** [comment]
Yes, putting a flag for outputs would make sense, or else, simple validations are needlesly spreaded in in 2 different places (as i see it, contracts are for higker level constraints that cannot jsonschema cannot easily express).

**Comment by ankostis** [comment]
Actually, revalidating schema on each node would be deadly slow.

What it could work would be 
a) to keep intermediate data-items and validate them at the end of the run, or
b)  hook into `jsonschema` library and evaluate just thre rules for each new item generated.

The later,which is better for a usability standpoint (it fails early & on the spot), unfortunately it is straight-forwards to implement only for trivial schemas.
